### PR TITLE
fix(clapi): insert poller into platform_topology

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description
This PR fix an issue where when using CLAPI to add a poller, it wasn't created into platform_topology.
**Fixes** # MON-10697

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.04.x
- [x] 20.10.x
- [x] 21.04.x
- [x] 21.10.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: lib/Centreon/Object/Instance/Instance.php [hunk 1]
```diff
@@ -56,4 +60,144 @@ public function getDefaultInstance()
...
+        if ($platformTopology !== null) {
+            if ($platformTopology->isPending() === false) {
+                throw new \Exception('Platform already created');
+            }
+
+            /**
+             * Check if the parent is a registered remote.
+             */
+            $parentPlatform = $this->findPlatformTopology($platformTopology->getParentId());
+            if ($parentPlatform !== null && $parentPlatform->getType() === 'remote') {
```
Comment (callapa) [inline lib/Centreon/Object/Instance/Instance.php:84]: Is there a constant to replace "remote"?

File: lib/Centreon/Object/Instance/Instance.php [hunk 2]
```diff
@@ -56,4 +60,144 @@ public function getDefaultInstance()
...
+
+    /**
+     * Insert platform in nagios_server and platform_topology tables.
+     *
+     * @param array<string,mixed> $params
+     * @return void
+     */
+    public function insert($params = [])
+    {
+        $platformTopology = $this->findPlatformTopologyByAddress($params['ns_ip_address']);
```
Comment (callapa) [inline lib/Centreon/Object/Instance/Instance.php:72]: Please test the existence of this 'ns_ip_address' index first.

File: lib/Centreon/Object/Instance/Instance.php [hunk 3]
```diff
@@ -56,4 +60,144 @@ public function getDefaultInstance()
...
+                $platformTopology->setPending(false);
+                $platformTopology->setServerId($serverId);
+                $this->updatePlatformTopology($platformTopology);
+                $this->db->commit();
+            } catch (\Exception $ex) {
+                $this->db->rollBack();
+                throw new \Exception('Unable to update platform', 0, $ex);
+            }
+        } else {
+            $this->db->beginTransaction();
```
Comment (callapa) [inline lib/Centreon/Object/Instance/Instance.php:103]: Please check that we are not already in a database transaction.

**Review by callapa — CHANGES_REQUESTED** [review]

File: lib/Centreon/Object/Instance/Instance.php [hunk 4]
```diff
@@ -56,4 +61,152 @@ public function getDefaultInstance()
...
+        } else {
+            $isAlreadyInTransaction = $this->db->inTransaction();
+            if (!$isAlreadyInTransaction) {
+                $this->db->beginTransaction();
+            }
+            try {
+                $serverId = parent::insert($params);
+                $params['server_id'] = $serverId;
+                $this->insertIntoPlatformTopology($params);
+                $this->db->commit();
```
Comment (callapa) [inline lib/Centreon/Object/Instance/Instance.php:117]: If you are already in a database transaction, the commit/rollback process should be handled by whoever started it.
```suggestion
                if (!$isAlreadyInTransaction) {
                    $this->db->commit();
                }
```

File: lib/Centreon/Object/Instance/Instance.php [hunk 5]
```diff
@@ -56,4 +61,152 @@ public function getDefaultInstance()
...
+            } catch (\Exception $ex) {
+                $this->db->rollBack();
```
Comment (callapa) [inline lib/Centreon/Object/Instance/Instance.php:119]:
```suggestion
                if (!$isAlreadyInTransaction) {
                    $this->db->rollBack();
                }
```

**Review by callapa — CHANGES_REQUESTED** [review]

File: lib/Centreon/Object/Instance/Instance.php [hunk 6]
```diff
@@ -56,4 +61,153 @@ public function getDefaultInstance()
...
+     */
+    private function insertIntoPlatformTopology(array $params): void
+    {
+        $centralPlatformTopologyId = $this->findCentralPlatformTopologyId();
+        if ($centralPlatformTopologyId === null) {
+            throw new \Exception('No Central found in topology');
+        }
+        $statement = $this->db->prepare(
+            "INSERT INTO platform_topology (address, name, type, pending, parent_id, server_id) " .
+            "VALUES (:address, :name, 'poller', '0', :parentId, :serverId)"
```
Comment (kduret) [inline lib/Centreon/Object/Instance/Instance.php:191]: use a constant for `poller`

**Review by kduret — DISMISSED** [review]

File: lib/Centreon/Object/Instance/Instance.php [hunk 7]
```diff
@@ -56,4 +61,155 @@ public function getDefaultInstance()
         $row = $res->fetch();
         return $row['name'];
     }
+
+    /**
+     * Insert platform in nagios_server and platform_topology tables.
+     *
+     * @param array<string,mixed> $params
+     * @return void
```
Comment (callapa) [inline lib/Centreon/Object/Instance/Instance.php:69]:
```suggestion
     * @return int
```

**Review by callapa — CHANGES_REQUESTED** [review]

**Review by callapa — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**